File: android/app/src/main/java/com/solace/ui/journal/JournalViewModel.kt
package com.solace.ui.journal

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solace.data.journal.JournalEntry
import com.solace.data.journal.JournalRepository
import com.solace.ui.config.MoodColorKey
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class JournalUiState(
    val entries: List<JournalEntry> = emptyList(),
    val isLoading: Boolean = true,
    val isSaving: Boolean = false,
    val showCompose: Boolean = false,
    val draftEmoji: String = "😐",
    val draftText: String = "",
    val draftMoodColor: String = MoodColorKey.CALM_RELAXED.name,
    val error: String? = null,
)

@HiltViewModel
class JournalViewModel @Inject constructor(
    private val repo: JournalRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(JournalUiState())
    val state: StateFlow<JournalUiState> = _state.asStateFlow()

    private var currentUid = ""

    fun init(uid: String) {
        if (currentUid == uid) return
        currentUid = uid
        viewModelScope.launch {
            repo.getEntries(uid).collect { entries ->
                _state.update { it.copy(entries = entries, isLoading = false) }
            }
        }
    }

    fun openCompose() = _state.update {
        it.copy(showCompose = true, draftEmoji = "😐", draftText = "")
    }

    fun dismissCompose() = _state.update { it.copy(showCompose = false) }

    fun setDraftEmoji(emoji: String) = _state.update { it.copy(draftEmoji = emoji) }
    fun setDraftText(text: String) = _state.update { it.copy(draftText = text) }

    fun saveEntry() {
        val s = _state.value
        if (s.draftText.isBlank()) return
        viewModelScope.launch {
            _state.update { it.copy(isSaving = true) }
            repo.addEntry(currentUid, s.draftEmoji, s.draftText.trim(), s.draftMoodColor)
                .onSuccess { _state.update { it.copy(showCompose = false, isSaving = false) } }
                .onFailure { e -> _state.update { it.copy(isSaving = false, error = e.message) } }
        }
    }

    fun deleteEntry(entryId: String) {
        viewModelScope.launch {
            repo.deleteEntry(currentUid, entryId)
        }
    }

    fun clearError() = _state.update { it.copy(error = null) }
}

File: ios/Solace/ViewModels/JournalViewModel.swift
import Foundation
import FirebaseFirestore

@MainActor
final class JournalViewModel: ObservableObject {

    @Published var entries: [JournalEntry] = []
    @Published var isLoading: Bool = true
    @Published var isSaving: Bool = false
    @Published var showCompose: Bool = false
    @Published var draftEmoji: String = "😐"
    @Published var draftText: String = ""
    @Published var error: String? = nil

    private var currentUid: String = ""
    private var listener: ListenerRegistration?
    private let repo = JournalRepository.shared

    func initialize(uid: String) {
        guard uid != currentUid else { return }
        currentUid = uid
        listener?.remove()
        listener = repo.observeEntries(uid: uid) { [weak self] entries in
            guard let self else { return }
            self.entries = entries
            self.isLoading = false
        }
    }

    func openCompose() {
        draftEmoji = "😐"
        draftText = ""
        showCompose = true
    }

    func dismissCompose() { showCompose = false }

    func saveEntry() {
        let text = draftText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        isSaving = true
        Task {
            do {
                _ = try await repo.addEntry(uid: currentUid, emoji: draftEmoji, text: text, moodColor: "CALM_RELAXED")
                showCompose = false
            } catch {
                self.error = error.localizedDescription
            }
            isSaving = false
        }
    }

    func deleteEntry(_ entry: JournalEntry) {
        Task {
            try? await repo.deleteEntry(uid: currentUid, entryId: entry.id)
        }
    }

    deinit { listener?.remove() }
}
